### PR TITLE
Mpls cli

### PR DESCRIPTION
**Pull request by devadityaojha**
#### What I did

Added cli show interface commands for mpls 

#### How I did it

Modified the logic for "show interfaces mpls" from single-asic vs in __init__.py under /show/interfaces to now accepting multi-asic vs commands. Also added a series of unit tests to test out functionality of the new commands. 

#### How to verify it

Run the following commands, will output mpls statuses on the respective interfaces
"show interfaces mpls" -> will only output frontend interfaces
"show interfaces mpls -n $ASIC#" -> will output all frontend interfaces on that asic
"show interfaces mpls -n $ASIC# -d all" -> will output all interfaces on that asic (including backend)
"show interfaces mpls -d all" -> will output all interfaces on the vs (all frontend + backend interfaces)

#### Previous command output (if the output of a command-line utility has changed)

Previously, each command needed to have been run inside the docker shell for that asic, with broken output (no indication of mpls enabled or disabled on that interface) 
Ex. "sudo ip netns exec asic0 show interfaces mpls"

Output:
Interface                        MPLS State
-----------                      --------------
Loopback0  
Loopback4096  
PortChannel4001     
PortChannel4002    

"show interfaces mpls" on the host machine also outputted nothing

#### New command output (if the output of a command-line utility has changed)

The previous command has now become 
"show interfaces mpls -n $ASIC#" (can be run directly on the host machine)

Ex. "show interfaces mpls -n asic0"

Output:
Interface                        MPLS State
-----------                      --------------

(fyi, no frontend interfaces, hence no output, as it correctly should be ^)

Ex. "show interfaces mpls -n asic0 -d all"

Output:
Interface                        MPLS State
-----------                      --------------
Loopback0                     disable
Loopback4096               disable
PortChannel4001           enable
PortChannel4002           enable

(^ Backend interfaces are also now listed correctly using the "-d all" flag)

"show interfaces mpls" outputs all frontend interfaces directly on the host machine

There are no frontend interfaces on my specific vs, so nothing is outputted. 

However, if I run "show interfaces mpls -d all" 

Output: 
Interface               MPLS State
---------------       ------------
Loopback0            disable
Loopback4096      disable
PortChannel4001  enable
PortChannel4002  enable
PortChannel4003  enable
PortChannel4004  enable
PortChannel4009  enable
PortChannel4010  enable
PortChannel4013  enable
PortChannel4014  enable

### CODE REVIEW

File: show/interfaces/__init__.py [hunk 1]
```diff
@@ -324,33 +324,60 @@ def expected(db, interfacename):
 # 'mpls' subcommand ("show interfaces mpls")
 @interfaces.command()
 @click.argument('interfacename', required=False)
+#@multi_asic_util.multi_asic_click_options
```
Comment (smaheshm) [inline show/interfaces/__init__.py:327]: why not use decorator?

Comment (devadityaojha) [reply show/interfaces/__init__.py:327]: I call it in the function.

File: tests/interfaces_test.py [hunk 2]
```diff
@@ -40,6 +40,7 @@
 Ethernet116  etp30
 Ethernet120  etp31
 Ethernet124  etp32
+Ethernet128  etp33
```
Comment (smaheshm) [inline tests/interfaces_test.py:43]: why add this new entry?

File: tests/mock_tables/config_db.json [hunk 3]
```diff
@@ -370,6 +370,16 @@
         "pfc_asym": "off",
         "speed": "40000"
     },
+    "PORT|Ethernet128": {
```
Comment (smaheshm) [inline tests/mock_tables/config_db.json:373]: I see why you added etp33, can't we use existing ports.

File: tests/interfaces_test.py (hunk 2 above)
Comment (devadityaojha) [reply tests/interfaces_test.py:43]: internal interface for single asic, none were present initially

File: tests/mpls_input/appl1_db.json [hunk 4]
```diff
@@ -0,0 +1,47 @@
+{
```
Comment (smaheshm) [inline tests/mpls_input/appl1_db.json:1]: are these files required? Can't we use the ones in "asic0/xx" and "asic1/xx"

File: tests/mpls_test.py [hunk 5]
```diff
@@ -1,4 +1,6 @@
 import os
+import importlib
+#from importlib import reload
```
Comment (smaheshm) [inline tests/mpls_test.py:3]: remove this line if not required.

File: tests/mpls_test.py [hunk 6]
```diff
@@ -23,71 +25,203 @@
...
 mock_db_path = os.path.join(test_path, "mpls_input")
+masic_mock_db_path = os.path.join(test_path, "mock_tables")
+masic0_mock_db_path = os.path.join(masic_mock_db_path, "asic0")
 
 
 class TestMpls(object):
     @classmethod
     def setup_class(cls):
         print("SETUP")
-        os.environ['UTILITIES_UNIT_TESTING'] = "1"
```
Comment (smaheshm) [inline tests/mpls_test.py:43]: why remove this line. Some scripts depend on the env variable to load mock APIs

File: tests/mpls_input/appl1_db.json (hunk 4 above)
Comment (devadityaojha) [reply tests/mpls_input/appl1_db.json:1]: no, will remove them

File: tests/mock_tables/config_db.json (hunk 3 above)
Comment (devadityaojha) [reply tests/mock_tables/config_db.json:373]: don't think so

File: tests/pfcwd_input/pfcwd_test_vectors.py [hunk 7]
```diff
@@ -44,11 +44,11 @@
 """
 
 pfcwd_show_start_default = """\
-Changed polling interval to 200ms
+Changed polling interval to 400ms
```
Comment (smaheshm) [inline tests/pfcwd_input/pfcwd_test_vectors.py:47]: why change this value?

File: tests/mpls_test.py (hunk 5 above)
Comment (devadityaojha) [reply tests/mpls_test.py:3]: ok

File: tests/mpls_test.py [hunk 8]
```diff
@@ -23,71 +25,203 @@
...
 
     @classmethod
     def teardown_class(cls):
         print("TEARDOWN")
         os.environ['UTILITIES_UNIT_TESTING'] = "0"
-        dbconnector.dedicated_dbs['APPL_DB'] = None
+        from .mock_tables import mock_single_asic
+        importlib.reload(mock_single_asic)
+        from .mock_tables import dbconnector
+        dbconnector.load_database_config
```
Comment (smaheshm) [inline tests/mpls_test.py:227]: dbconnector.load_database_config()?

File: tests/mpls_test.py (hunk 6 above)
Comment (devadityaojha) [reply tests/mpls_test.py:43]: sure

File: tests/pfcwd_input/pfcwd_test_vectors.py (hunk 7 above)
Comment (devadityaojha) [reply tests/pfcwd_input/pfcwd_test_vectors.py:47]: the corresponding output changed on the test

File: tests/mpls_test.py (hunk 8 above)
Comment (devadityaojha) [reply tests/mpls_test.py:227]: will correct

File: show/main.py [hunk 9]
```diff
@@ -15,6 +16,25 @@
...
 
+# mock the redis for unit test purposes #
+try:
+    if os.environ["UTILITIES_UNIT_TESTING"] == "2":
+        modules_path = os.path.join(os.path.dirname(__file__), "..")
+        tests_path = os.path.join(modules_path, "tests")
+        sys.path.insert(0, modules_path)
+        sys.path.insert(0, tests_path)
+        import mock_tables.dbconnector
+        print(" here ")
```
Comment (smaheshm) [inline show/main.py:27]: remove

File: tests/interfaces_test.py (hunk 2 above)
Comment (smaheshm) [reply tests/interfaces_test.py:43]: there is no interface interfaces in single ASIC

File: tests/intfutil_test.py [hunk 10]
```diff
@@ -160,14 +160,14 @@ def test_show_interfaces_description_etp9_in_alias_mode(self):
         assert result.exit_code == 0
         assert result.output == show_interface_description_eth9_output
 
-    def test_show_interfaces_description_etp33_in_alias_mode(self):
+    def test_show_interfaces_description_etp34_in_alias_mode(self):
```
Comment (smaheshm) [inline tests/intfutil_test.py:163]: Once the 'internal interface' is removed for single ASIC these changes will not be required.

File: tests/intfutil_test.py [hunk 11]
```diff
@@ -200,47 +199,13 @@ def test_subintf_status_verbose(self):
         expected_output = "Command: intfutil -c status -i subport"
         self.assertEqual(result.output.split('\n')[0], expected_output)
 
-
```
Comment (smaheshm) [inline tests/intfutil_test.py:203]: also bring back these tests.

File: tests/mock_tables/appl_db.json [hunk 12]
```diff
@@ -135,8 +135,17 @@
         "role": "Int",
         "speed": "100000"
     },
-    "INTF_TABLE:Ethernet0.10": {
-        "admin_status": "up"
+    "INTF_TABLE:Ethernet16": {
+        "mpls": "enable"
+    },
```
Comment (smaheshm) [inline tests/mock_tables/appl_db.json:140]: are these changes required since you are adding tests only for multi ASIC

File: tests/mpls_input/appl_db.json [hunk 13]
```diff
@@ -1,47 +0,0 @@
-{
```
Comment (smaheshm) [inline tests/mpls_input/appl_db.json:1]: this one too.. is it required to removed? Does it affect multi ASIC test cases?

File: tests/mpls_test.py [hunk 14]
```diff
@@ -23,71 +24,201 @@
...
+        print(result.output)
+        assert result.exit_code == 0
+        assert result.output == show_interfaces_mpls_masic_output_asic_all
+    
+    def test_config_mpls_masic_remove(self):
+        runner = CliRunner()
+        db = Db()
+        obj = {'config_db':db.cfgdb, 'namespace':'asic0'}
+
+        result = runner.invoke(config.config.commands["interface"].commands["mpls"].commands["remove"], ["Ethernet8"], obj=obj)
```
Comment (smaheshm) [inline tests/mpls_test.py:211]:
```
result = runner.invoke(
    config.config.commands["interface"].commands["mpls"].commands["remove"],
    ["Ethernet8"], obj=obj
)
```

File: tests/mpls_test.py [hunk 15]
```diff
@@ -23,71 +24,201 @@
...
         print(result.exit_code)
         print(result.output)
         assert result.exit_code == 0
-        assert result.output == show_interfaces_mpls_specific_output
+        assert db.cfgdb.get_entry("INTERFACE", "Ethernet8") == {"mpls": "disable"}
```
Comment (smaheshm) [inline tests/mpls_test.py:216]: How about adding a negative test case for namespace. 'i.e' pass an invalid namespace in the option "-ndummy". This way  you can test "multi_asic.validation" callback.

File: show/interfaces/__init__.py (hunk 1 above)
Comment (devadityaojha) [reply show/interfaces/__init__.py]: resolved

File: tests/interfaces_test.py (hunk 2 above)
Comment (devadityaojha) [reply tests/interfaces_test.py:43]: will remove

File: show/main.py (hunk 9 above)
Comment (devadityaojha) [reply show/main.py:27]: ok

File: show/interfaces/__init__.py [hunk 16]
```diff
@@ -321,36 +321,61 @@ def expected(db, interfacename):
 
     click.echo(tabulate(body, header))
 
-# 'mpls' subcommand ("show interfaces mpls")
 @interfaces.command()
 @click.argument('interfacename', required=False)
+@click.option('--namespace',
```
[the remark this answers is not in the document]
Comment (devadityaojha) [reply show/interfaces/__init__.py:326]: ok

File: tests/intfutil_test.py (hunk 10 above)
Comment (devadityaojha) [reply tests/intfutil_test.py:163]: yup

File: tests/intfutil_test.py (hunk 11 above)
Comment (devadityaojha) [reply tests/intfutil_test.py:203]: ok

File: tests/mpls_input/appl_db.json (hunk 13 above)
Comment (devadityaojha) [reply tests/mpls_input/appl_db.json:1]: no, will bring back

File: tests/mpls_test.py (hunk 15 above)
Comment (devadityaojha) [reply tests/mpls_test.py:216]: unfortunately not possible to test, as variables cannot be modified

File: tests/mock_tables/appl_db.json (hunk 12 above)
Comment (devadityaojha) [reply tests/mock_tables/appl_db.json:140]: reverted

File: tests/mpls_input/appl_db.json [hunk 17]
```diff
@@ -26,22 +23,10 @@
...
-    },
-    "INTF_TABLE:PortChannel2": {
+    "INTF_TABLE:Ethernet12": {
         "mpls": "disable"
     },
-    "INTF_TABLE:PortChannel2:10.0.0.56/31": {
-        "NULL": "NULL"
-    },
-    "INTF_TABLE:Vlan2": {
+    "INTF_TABLE:Ethernet20": {
```
Comment (devadityaojha) [inline tests/mpls_input/appl_db.json:29]: In case you're wondering why some of the ports from the previous version do not appear here, it is because those ports do not appear in mock_tables/configdb

**Review by smaheshm — APPROVED** [review]
looks good.
